src/ast.rs

```diff
@@ -510,24 +510,36 @@ impl Tag {
                 };
                 for (key, value) in &mut object.fields {
                     // Infer whether the value is nullable
-                    value.infer_nullability();
+                    value.infer_nullability(enforce_nullable);
                     // A required nullable field is still nullable
                     value.nullable |= !required.contains(key);
+                    // All fields are nullable if enforced
+                    if enforce_nullable {
+                        value.nullable = true;
+                    }
                 }
             }
-            Type::Map(map) => map.value.infer_nullability(),
-            Type::Array(array) => array.items.infer_nullability(),
+            Type::Map(map) => map.value.infer_nullability(enforce_nullable),
+            Type::Array(array) => array.items.infer_nullability(enforce_nullable),
             Type::Tuple(tuple) => {
                 for item in tuple.items.iter_mut() {
-                    item.infer_nullability();
+                    item.infer_nullability(enforce_nullable);
                 }
             }
             Type::Union(union) => {
                 for item in union.items.iter_mut() {
-                    item.infer_nullability();
+                    item.infer_nullability(enforce_nullable);
                 }
             }
-            _ => (),
+            _ => {
+                // Other types, like atoms, are set to nullable if enforced
+                if enforce_nullable {
+                    self.nullable = true;
+                }
+            }
+        }
+        if enforce_nullable {
+            self.nullable = true
```

Comment:
I don't understand why this is setting `self.nullable = true` both here and on 537

Reply:
This is necessary, otherwise the top-level object won't be set to nullable. This case was caught by a unit test in line 1254.

Reply:
but then why is the one on 537 necessary?

Reply:
The fall through case where the type is atomic. I see the redundancy now though.